inst/template/interface.cpp

```diff
@@ -0,0 +1,90 @@
+#include <dust/dust.hpp>
+#include <dust/util.hpp>
+
+{{model}}
+
+extern "C" void {{name}}_finalise(SEXP ptr) {
+  dust::Dust<{{type}}> *obj =
+    dust::util::read_r_pointer<dust::Dust<{{type}}>>(ptr, false);
+  if (obj) {
+    delete obj;
+  }
+  if (ptr) {
+    R_ClearExternalPtr(ptr);
+  }
+}
+
+extern "C" SEXP {{name}}_alloc(SEXP data, SEXP r_step,
+                               SEXP r_n_particles, SEXP r_n_threads,
+                               SEXP r_n_generators, SEXP r_seed) {
+  size_t step = dust::util::as_size(r_step, "step");
+  size_t n_particles = dust::util::as_size(r_n_particles, "n_particles");
+  size_t n_threads = dust::util::as_size(r_n_threads, "n_threads");
+  size_t n_generators = dust::util::as_size(r_n_generators, "n_generators");
+  size_t seed = dust::util::as_size(r_seed, "seed");
+  dust::util::validate_n(n_generators, n_threads);
+
+  std::vector<size_t> index_y = {0};
+
+  dust::Dust<{{type}}> *d =
```

Comment:
Would wrapping this in a `shared_ptr` work, rather than `new` and `delete` in the finaliser? Or is that not easily compatible with the R interface?

Reply:
Obsolete now that we're using Rcpp